mobile_verifier/src/sp_boosted_rewards_bans.rs

```diff
@@ -0,0 +1,654 @@
+        &self,
+        transaction: &mut Transaction<'_, Postgres>,
+        ingest: ServiceProviderBoostedRewardsBannedRadioIngestReportV1,
+    ) -> anyhow::Result<()> {
+        let report = BannedRadioReport::try_from(ingest.clone())?;
+        let is_authorized = self.is_authorized(&report.pubkey).await?;
+
+        if is_authorized {
+            db::update_report(transaction, &report).await?;
+            self.update_seniority(transaction, &report).await?;
+        }
+
+        let status = match is_authorized {
+            true => ServiceProviderBoostedRewardsBannedRadioVerificationStatus::SpBoostedRewardsBanValid,
+            false => ServiceProviderBoostedRewardsBannedRadioVerificationStatus::SpBoostedRewardsBanInvalidCarrierKey,
+        };
+
+        let verified_report = VerifiedServiceProviderBoostedRewardsBannedRadioIngestReportV1 {
+            report: Some(ingest),
+            status: status.into(),
+            timestamp: Utc::now().timestamp_millis() as u64,
+        };
+
+        self.verified_sink
+            .write(verified_report, &[("status", status.as_str_name())])
+            .await?;
+
+        Ok(())
+    }
+
+    async fn is_authorized(&self, pubkey: &PublicKeyBinary) -> anyhow::Result<bool> {
+        self.authorization_verifier
+            .verify_authorized_key(pubkey, NetworkKeyRole::MobileCarrier)
+            .await
+            .map_err(anyhow::Error::from)
+    }
+
+    async fn update_seniority(
+        &self,
+        transaction: &mut Transaction<'_, Postgres>,
+        report: &BannedRadioReport,
+    ) -> anyhow::Result<()> {
+        if let Some(current_seniority) =
+            Seniority::fetch_latest(report.key.to_ref(), transaction).await?
+        {
+            let seniority_update = SeniorityUpdate::new(
+                report.key.to_ref(),
+                current_seniority.last_heartbeat,
+                current_seniority.uuid,
+                SeniorityUpdateAction::Insert {
+                    new_seniority: Utc::now(),
+                    update_reason: SeniorityUpdateReason::ServiceProviderBan,
+                },
+            );
+
+            seniority_update.write(&self.seniority_update_sink).await?;
+            seniority_update.execute(transaction).await?;
+        }
+
+        Ok(())
+    }
+}
+
+pub async fn clear_bans(
+    transaction: &mut Transaction<'_, Postgres>,
+    before: DateTime<Utc>,
+) -> anyhow::Result<()> {
+    db::cleanup(transaction, before).await
+}
+
+pub mod db {
+    use std::str::FromStr;
+
+    use chrono::Duration;
+    use sqlx::Row;
+
+    use super::*;
+
+    pub async fn get_banned_radios(
+        pool: &PgPool,
+        date_time: DateTime<Utc>,
+    ) -> anyhow::Result<BannedRadios> {
+        sqlx::query(
+            r#"
+                SELECT radio_type, radio_key
+                FROM sp_boosted_rewards_bans
+                WHERE until > $1 
+                    AND (invalidated_at > $1 OR invalidated_at IS NULL)
```

Comment:
you can use coalesce to simplify this: 
```suggestion
                    AND COALESCE(invalidated_at > $1, TRUE)
```

Reply:
addressed 15874cb